### PR TITLE
Initial Training

### PR DESCRIPTION
**Pull request by mek604**
The model now run after a bit of fixing. I made two versions of UNet: one with keras + tensorflow, the other is purely tensorflow v1.

### CODE REVIEW

**Comment by mek604** [comment]
I couldn't run the program on my machine, so I used Google CoLab. Still took a while to run on two inputs, one from DSD100_Npz and the other from MedleyDB_Npz. 

Here is the output:

```
(486, 512, 128, 1) (486, 512, 128, 1)
INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': './model', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_session_creation_timeout_secs': 7200, '_service': None, '_cluster_spec': ClusterSpec({}), '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow_estimator/python/estimator/inputs/queues/feeding_queue_runner.py:65: QueueRunner.__init__ (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow_estimator/python/estimator/inputs/queues/feeding_functions.py:491: add_queue_runner (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
INFO:tensorflow:Calling model_fn.
called
WARNING:tensorflow:From <ipython-input-14-09aa8e221d13>:12: conv2d (from tensorflow.python.keras.legacy_tf_layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.keras.layers.Conv2D` instead.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/legacy_tf_layers/convolutional.py:424: Layer.apply (from tensorflow.python.keras.engine.base_layer_v1) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
WARNING:tensorflow:From <ipython-input-14-09aa8e221d13>:37: conv2d_transpose (from tensorflow.python.keras.legacy_tf_layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.keras.layers.Conv2DTranspose` instead.
WARNING:tensorflow:From <ipython-input-14-09aa8e221d13>:38: dropout (from tensorflow.python.keras.legacy_tf_layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dropout instead.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from ./model/model.ckpt-31
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/saver.py:1077: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/monitored_session.py:906: start_queue_runners (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 31...
INFO:tensorflow:Saving checkpoints for 31 into ./model/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 31...
INFO:tensorflow:loss = 0.0027773918, step = 31
INFO:tensorflow:global_step/sec: 0.432653
INFO:tensorflow:loss = 0.0030719403, step = 131 (231.139 sec)
INFO:tensorflow:global_step/sec: 0.432179
INFO:tensorflow:loss = 0.003370784, step = 231 (231.382 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 291...
INFO:tensorflow:Saving checkpoints for 291 into ./model/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 291...
INFO:tensorflow:global_step/sec: 0.430518
INFO:tensorflow:loss = 0.0034812419, step = 331 (232.281 sec)
INFO:tensorflow:global_step/sec: 0.432727
INFO:tensorflow:loss = 0.0022151503, step = 431 (231.090 sec)
INFO:tensorflow:global_step/sec: 0.432261
INFO:tensorflow:loss = 0.002953046, step = 531 (231.341 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 551...
INFO:tensorflow:Saving checkpoints for 551 into ./model/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 551...
INFO:tensorflow:global_step/sec: 0.428604
INFO:tensorflow:loss = 0.001916614, step = 631 (233.319 sec)
INFO:tensorflow:global_step/sec: 0.432919
INFO:tensorflow:loss = 0.0018378615, step = 731 (230.987 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 811...
INFO:tensorflow:Saving checkpoints for 811 into ./model/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 811...
INFO:tensorflow:global_step/sec: 0.433236
INFO:tensorflow:loss = 0.0014427314, step = 831 (230.821 sec)
INFO:tensorflow:global_step/sec: 0.435925
INFO:tensorflow:loss = 0.0024421935, step = 931 (229.400 sec)
INFO:tensorflow:global_step/sec: 0.435144
INFO:tensorflow:loss = 0.002258926, step = 1031 (229.806 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 1072...
INFO:tensorflow:Saving checkpoints for 1072 into ./model/model.ckpt.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/training/saver.py:971: remove_checkpoint (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to delete files with this prefix.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 1072...
INFO:tensorflow:global_step/sec: 0.4316
INFO:tensorflow:loss = 0.001930545, step = 1131 (231.699 sec)
INFO:tensorflow:global_step/sec: 0.436063
INFO:tensorflow:loss = 0.002292505, step = 1231 (229.324 sec)
INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 1246...
INFO:tensorflow:Saving checkpoints for 1246 into ./model/model.ckpt.
INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 1246...
INFO:tensorflow:Loss for final step: 0.002101112.

<tensorflow_estimator.python.estimator.estimator.EstimatorV2 at 0x7fc5f65ec2b0>
```

**Comment by mek604** [comment]
Good lord, that was 42 minutes

File: src/unet.py [hunk 1]
```diff
@@ -1,76 +1,79 @@
...
+    input_layer = tf.reshape(features['mix'],input_shape)
+
     # # Convolutional Layer 1
-    # conv1 = tf.keras.layers.batch_normalization(tf.keras.layers.conv2d(inputs = input_layer, filters=16, kernel_size=[5,5], 
-    #                                                    strides=[2,2], padding="same", activation=tf.nn.leaky_relu))
-    
-    # # Convolutional Layer 2 
-    # conv2 = tf.layers.batch_normalization(tf.layers.conv2d(inputs = conv1, filters = 32, kernel_size = [5,5], 
-    #                                                    strides = [2,2], padding="same", activation = tf.nn.leaky_relu))
+    conv1 = BatchNormalization()(tf.keras.layers.Conv2D(filters=16, kernel_size=[5,5],
```
Comment (eyvorchuk) [inline src/unet.py:12]: Change this to just `Conv2D(...)` rather than `tf.keras.layers.Conv2D(...)` to be consistent with the other layers.

**Review by eyvorchuk — APPROVED** [review]
This looks great! See what the others think, but I have a small suggestion.

File: src/unet.py (hunk 1 above)
Comment (mek604) [reply src/unet.py]: completed

**Comment by mek604** [comment]
The dropout layers in the main branch have the training mode argument checked:
```
dropout2 = tf.layers.dropout(inputs = deconv2, rate = 0.5, training=mode == tf.estimator.ModeKeys.TRAIN)
```
This argument is incompatible with keras dropout, so if we need to keep it I will have to change

```dropout2 = Dropout(rate = 0.5)(deconv2)``` to

 `` tf.compat.v1.layers.dropout(inputs = deconv2, rate = 0.5, training=mode == tf.estimator.ModeKeys.TRAIN)``

File: src/unet_tf_v1 [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                                             kernel_size=[5, 5],
+                                             strides=[2, 2], padding="same", activation=tf.nn.relu))
+    # Deconvolutional Layer6
+    deconv6 = tf.compat.v1.layers.conv2d_transpose(inputs=tf.compat.v1.concat([deconv5, conv1], 3), filters=1,
+                                                   kernel_size=[5, 5],
+                                                   strides=[2, 2], padding="same", activation=tf.nn.relu)
+
+    predictions = {'outputs': deconv6
+                   }
+
```
Comment (mek604) [inline src/unet_tf_v1:76]: The final layer in the paper uses  __sigmoid__ activation function. Does that mean deconv6 is the mask layer, which is also the final layer?

**Comment by mek604** [comment]
Here is the output of running the model with all input files
[output.txt](https://github.com/Zhz1997/Singing-voice-speration-with-U-Net/files/5520346/output.txt)
